### PR TITLE
A Logger

### PR DESCRIPTION
**Pull request by harrynull**
This pull request implemented a stream-based, extensible Logger that differs outputs of different levels.
* It's extensible: It uses sinks to manage where it outputs to. You can set different minimum output level for each sink. Everything inheriting std::ostream can be a sink, like a file, or even a network server.
* It's also easy to use: you don't need to specify the types of the variable like %d, %s. It can automatically print the variable. Also, it is no longer limited to only the built-in types.
* Also, it's colorful :)
## Usage
To initialize:
```cpp
Logger::Sink sink(logFile, false); // logFile is a std::ofstrea;, false means don't use color for this sink.
sink.setMinimumOutputLevel(Logger::Level::verbose); // forward all logs to the log file.
getLogger().addSink(sink); // add a log file. (std::cout is automatically added)
```
To log:
```cpp
debugstream << "Creating temporary directories at tempFiles/";
infostream << "Grammar files created!";
fatalstream(EXIT_INVALID_PARAMETERS) << "Unsupported Aligner Type!";
```
## Screenshots
linux_preview
linux_log_saved
linux_colors
win_levels
win_1
win_2
use_verbosity

### CODE REVIEW

File: src/lib_ccaligner/phoneme_utils.cpp [hunk 1]
```diff
@@ -120,7 +120,7 @@ std::vector<Phoneme> stringToPhoneme(const std::string &word)
         Phoneme phone = charToPhone(c);
         if (phone == "Noise")
         {
-            std::cout<<"Error converting "<<phone<<"\n";
+            LOG("Error converting %s", phone.c_str());
```
Comment (saurabhshri) [inline src/lib_ccaligner/phoneme_utils.cpp]: 👍

File: src/lib_ccaligner/phoneme_utils.h [hunk 2]
```diff
@@ -16,7 +16,7 @@
 #define CCALIGNER_PHONEME_UTILS_H
 
 #include "commons.h"
-typedef std::string Phoneme;
+using Phoneme = std::string;
```
Comment (saurabhshri) [inline src/lib_ccaligner/phoneme_utils.h]: Don't they mean the exact same thing? :) But yeah, this looks cleaner.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 3]
```diff
@@ -33,8 +33,8 @@ PocketsphinxAligner::PocketsphinxAligner(Params* parameters) noexcept
 {
     LOG("Initialising Aligner using PocketSphinx");
     LOG("Audio Filename: %s Subtitle filename: %s", _audioFileName.c_str(), _subtitleFileName.c_str());
-
-    std::cout << "Reading and decoding audio samples...\n";
+        
+    LOG("Reading and decoding audio samples...");
```
Comment (saurabhshri) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: Intentional.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 4]
```diff
@@ -49,12 +49,12 @@ bool PocketsphinxAligner::generateGrammar(grammarName name)
 {
     LOG("Generating Grammar based on subtitles, Grammar Name: %d ", name);
 
-    std::cout << "Generating language model and grammar files...\n";
+	LOG("Generating language model and grammar files...");
```
Comment (saurabhshri) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: Intentional.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 5]
```diff
@@ -49,12 +49,12 @@ bool PocketsphinxAligner::generateGrammar(grammarName name)
...
 
     if(_parameters->grammarType == complete_grammar || _parameters->grammarType == dict)
     {
-        std::cout << "Note: You have chosen to generate a dictionary. Based on your TensorFlow configuration,\n";
-        std::cout << "this may take some time, please be patient. For alternatives, see docs.\n";
+		LOG("Note: You have chosen to generate a dictionary. Based on your TensorFlow configuration,");
+		LOG("this may take some time, please be patient. For alternatives, see docs.");
```
Comment (saurabhshri) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: Intentional.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 6]
```diff
@@ -427,7 +424,7 @@ bool PocketsphinxAligner::recognise()
         recognitionWindow = _sampleWindow;
     }
 
-    std::cout << "Recognising and aligning..\n";
+    LOG("Recognising and aligning..");
```
Comment (saurabhshri) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: Intentional.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 7]
```diff
@@ -525,14 +522,14 @@ bool PocketsphinxAligner::recognise()
             case karaoke:   subCount = printKaraokeContinuous(_outputFileName, subCount, sub, _parameters->printOption);
                 break;
 
-            default:        std::cout<<"An error occurred while choosing output format!";
+            default:        FATAL(EXIT_UNKNOWN, "An error occurred while choosing output format!");
                 exit(2);
```
Comment (saurabhshri) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: This exit statement could be deleted. Maybe `EXIT_INVALID_PARAMETERS` fits more better here as well.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 8]
```diff
@@ -525,14 +522,14 @@ bool PocketsphinxAligner::recognise()
...
         }
     }
 
     printFileEnd(_outputFileName, _parameters->outputFormat);
 
-    std::cout << "Finished recognition and alignment..\n";
+    LOG("Finished recognition and alignment..");
```
Comment (saurabhshri) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: Intentional.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 9]
```diff
@@ -626,7 +623,7 @@ int PocketsphinxAligner::findTranscribedWordTimings(cmd_ln_t *config, ps_decoder
 
 bool PocketsphinxAligner::transcribe()
 {
-    std::cout << "Transcribing...\n";
+    LOG("Transcribing...");
```
Comment (saurabhshri) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: Intentional.

File: src/lib_ccaligner/recognize_using_pocketsphinx.cpp [hunk 10]
```diff
@@ -694,7 +691,7 @@ bool PocketsphinxAligner::transcribe()
 
     printTranscriptionFooter(_outputFileName, _parameters->outputFormat);
 
-    std::cout << "Finished transcription.\n";
+    LOG("Finished transcription.");
```
Comment (saurabhshri) [inline src/lib_ccaligner/recognize_using_pocketsphinx.cpp]: Intentional.

**Comment by saurabhshri** [comment]
Looks good, just a couple of changes.

File: src/lib_ccaligner/grammar_tools.cpp [hunk 11]
```diff
@@ -73,7 +73,7 @@ bool generate(std::vector <SubtitleItem*> subtitles, grammarName name)
 
     if(name == corpus || name == complete_grammar)
     {
-        std::cout<<"Creating Corpus : tempFiles/corpus/corpus.txt\n";
+		LOG("Creating Corpus : tempFiles/corpus/corpus.txt");
```
[the remark this answers is not in the document]
Comment (harrynull) [reply src/lib_ccaligner/grammar_tools.cpp]: Personally, I think users do not need to be aware of these datiled steps as they usually don't understand them anyway. I recommend using different levels of logging to unify them yet differ them from other logs, like verbose/debug/info/warning/error/fatal.

[the remark this answers is not in the document]
Comment (saurabhshri) [reply src/lib_ccaligner/grammar_tools.cpp]: @harrynull That sounds good. You can go ahead and implement that :)

[the remark this answers is not in the document]
Comment (saurabhshri) [reply src/lib_ccaligner/grammar_tools.cpp]: @harrynull BTW, that's not an assumption that is ours to make :) We have parameter options to manipulate these so why not. But your several levels of logging idea sounds perfect.

[the remark this answers is not in the document]
Comment (harrynull) [reply src/lib_ccaligner/grammar_tools.cpp]: Do you mind if I implement it as stream? I think would be easier to use.
Like this:
```
DEBUG<<"Some log"<<someVariable;
```

**Comment by saurabhshri** [comment]
@harrynull please read this comment on @Rakete1111 's PR : https://github.com/saurabhshri/CCAligner/pull/43#issuecomment-349572805

File: src/lib_ccaligner/phoneme_utils.h (hunk 2 above)
Comment (harrynull) [reply src/lib_ccaligner/phoneme_utils.h]: They are the same, and this looks cleaner as you said :)

File: docs/usage_parameters.adoc [hunk 12]
```diff
@@ -29,11 +29,7 @@ _E.g.: ``ccaligner -raw tbbt.raw -srt tbbt.srt``_
 
 _E.g.: ``ccaligner -wav tbbt.wav -srt tbbt.srt``_
 
-|`-txt`
-|`/path/to/text/file`
-|Provide path to transcript file in txt format. Only transcription works with this parameter. (See: -transcribe)
-
-_E.g.: ``ccaligner -wav tbbt.wav -txt tbbt.txt``_
```
Comment (saurabhshri) [inline docs/usage_parameters.adoc]: Why is this deleted? It's a parameter added just recently.

**Comment by saurabhshri** [comment]
Please add the newly added function description with usage examples in the documents. They are under the `docs` directory. Feel free to create a new file specifically for logging.

Also, please resolve the conflicts to make PR merge ready.

**Comment by saurabhshri** [comment]
@harrynull the build is failing.

File: src/lib_ccaligner/read_wav_file.cpp [hunk 13]
```diff
@@ -344,14 +342,12 @@ int WaveFileData::processStreamHeader()
             }
         }
     }
-    LOG("Error occurred while processing stream header!");
-    FATAL(EXIT_UNKNOWN, "Error occurred while processing stream header!");
-    return -1;  //some error; more robust exit errors coming soon
+    FATAL(UnknownError) << "Error occurred while processing stream header!";
```
Comment (saurabhshri) [inline src/lib_ccaligner/read_wav_file.cpp]:
```
/Users/travis/build/saurabhshri/CCAligner/src/lib_ccaligner/read_wav_file.cpp:346:1: warning: control may reach end of non-void function [-Wreturn-type]
}
```
^
/Users/travis/build/saurabhshri/CCAligner/src/lib_ccaligner/read_wav_file.cpp:498:1: warning: control may reach end of non-void function [-Wreturn-type]
}

Comment (harrynull) [reply src/lib_ccaligner/read_wav_file.cpp]: FATAL will abort the program (there's not catch block) it won't reach the end of the function. I believe it's a false positive.

**Comment by saurabhshri** [comment]
@harrynull Awesome job! :) I have changed the title, hope you don't mind. Thanks!
